pennylane/tape/functions/unified.py

```diff
@@ -0,0 +1,210 @@
+    >>> type(y)
+    <class 'pennylane.tape.functions.tf.TensorFlowTensor'>
+
+    The original tensor is available via the ``data`` attribute.
+
+    In addition, this class defines various abstract methods that all subclasses
+    must define. These methods allow for common manipulations and
+    linear algebra transformations without the need for importing.
+
+    >>> y.ones_like()
+    tf.Tensor([1. 1. 1.], shape=(3,), dtype=float32)
+
+    Unless specified, the returned tensors are also unified tensors, allowing
+    for method chaining:
+
+    >>> y.ones_like().expand_dims(0)
+    tf.Tensor([[1. 1. 1.]], shape=(1, 3), dtype=float32)
+    """
+
+    def __new__(cls, tensor):
+        if isinstance(tensor, UnifiedTensor):
+            return tensor
+
+        if cls is not UnifiedTensor:
+            return super(UnifiedTensor, cls).__new__(cls)
+
+        if isinstance(tensor, (list, tuple)):
+            from .autograd import AutogradTensor, np
+
+            return AutogradTensor.__new__(AutogradTensor, np.array(tensor))
+
+        namespace = tensor.__class__.__module__.split(".")[0]
+
+        if namespace in ("pennylane", "autograd", "numpy"):
+            from .autograd import AutogradTensor
+
+            return AutogradTensor.__new__(AutogradTensor, tensor)
+
+        if namespace == "tensorflow":
+            from .tf import TensorFlowTensor
+
+            return TensorFlowTensor.__new__(TensorFlowTensor, tensor)
+
+        if namespace == "torch":
+            from .torch import TorchTensor
+
+            return TorchTensor.__new__(TorchTensor, tensor)
+
+        raise ValueError(f"Unknown tensor type {type(tensor)}")
+
+    def __init__(self, tensor):
+        self.data = tensor
+
+    def __repr__(self):
+        return self.data.__repr__()
+
+    def __str__(self):
+        return self.data.__str__()
+
+    def __len__(self):
+        return len(self.data)
+
+    def __mul__(self, other):
+        if isinstance(other, UnifiedTensor):
+            other = other.data
+
+        return self.__class__(self.data * other)
+
+    def __rmul__(self, other):
+        if isinstance(other, UnifiedTensor):
+            other = other.data
+
+        return self.__class__(other * self.data)
+
+    @staticmethod
+    def unwrap(tensors):
+        """Unwraps a list of tensors, converting any :class:`UnifiedTensor`
+        objects in the list into raw interface tensors.
+
+        Args:
+            tensors (list[array_like]): list of arrays, tensors, or :class:`~.UnifiedTensor` objects
+
+        Returns
+            list[array_like]: the input list with all :class:`UnifiedTensor` objects
+            unwrapped.
+
+        **Example**
+
+        >>> x = tf.Variable([0.4, 0.1, 0.5])
+        >>> y = UnifiedTensor(x)
+        >>> z = tf.constant([0.1, 0.2])
+
+        Note that this is a static method, so we must pass the unified tensors itself
+        if we would like it to be included.
+
+        >>> res = y.unwrap([y, z])

```

Comment:
So it was not better to make it a non-static method and call it like `[a.unwrap() for a in lst]`?

Reply:
Doing it this way avoids the list comprehension, which makes it slightly more efficient. It also allows for unwrapping lists of mixed type (UnifiedTensor _or_ regular arrays)